README.rst
==================
Django Redisession
==================
A Redis-based Django session engine for django.contrib.sessions.

This engine supports storing session data as a Redis String, or distributing sessions to Redis hashs for more efficient usage of Redis storage(without much extra CPU cost on Redis server. Refs http://antirez.com/post/redis-weekly-update-7.html and http://redis.io/topics/memory-optimization , you may need to tweak Redis conf for actual usage. The tech is referred as hash mode below).

Installation
============
#. Require Django version >= 1.3
#. ``pip install django-redisession`` or in source ``python setup.py install``.
#. Add `redisession` to INSTALLED_APPS in your current settings file.
#. Set ``SESSION_ENGINE='redisession.backend'`` in the settings file.
    
Configuration
=============
Configuration options to the django-redisession can be placed in dict `REDIS_SESSION_CONFIG`:
    - `SERVER` -- If set to a dict then the value will be used as arguments for redis.Redis() to make a redis connection instance. If set to a string then the django-redisession will get redisession.helper.get_redis to lookup the value in `REDIS_CONFIG` for existing connection instance, see below.
    - `USE_HASH` -- If set to True (hash mode), the django-redisession will store sessions as items of Redis Hashes. `HASH_KEY_GENERATOR` will be used to generate corresponding Redis keys. If set to False, the data of a session will be stored as Redis String value of the Redis key which is generated by `KEY_GENERATOR`. session keys generated by md5 is not as well distribudated as those by sha1, but still acceptable.
    - `KEY_GENERATOR` -- The callable to generate Redis key, or Redis Hash field name when in hash mode, for a session. The callable accepts session_key as the only parameter. The default generates the binary string, 16 bytes for md5, from session_key which is normally a hex string. **Note** if you are customizing it, open `LOG_KEY_ERROR` and run test `./manage.py test redisession` to "ensure" `KEY_GENERATOR` works with session keys. For session key contains '0-9a-z' in incoming Django 1.5, you could check comments in `redisession/backend.py`. 
    - `HASH_KEY_GENERATOR` -- The callable to generate Redis key for Redis Hashes, in hash mode. The default uses the first two bytes of the binary string of a session_key, and thus distributes sessions to 65536 hashes. Other dispatching logics or key prefix could be achieved by customizing this option. **Note** if you are customizing it, open `LOG_KEY_ERROR` and run test `./manage.py test redisession` to "ensure" `HASH_KEY_GENERATOR` works with session keys. For session key contains '0-9a-z' in incoming Django 1.5, you could check comments in `redisession/backend.py`. 
    - `HASH_KEYS_CHECK_FOR_EXPIRY` -- The callable to generate Redis keys to check for deleting expired sessions, in hash mode. The callable accepts redis connection instance as parameter. The default applys Redis.randomkey 100 times to sample keys. 
    - `COMPRESS_LIB` -- The module name of the compress library used to compress session data, leave blank to disable compress. Defaults to 'snappy', refs `snappy <http://code.google.com/p/snappy/>`_ and `python-snappy <https://github.com/andrix/python-snappy>`_ . The compress library should support compress and decompress, such as zlib, lzo, etc. A slow library may have higher compression ratio, which implies higher load on web server, but less IO on communicating w/ Redis. So decide it by your tradeoff.
    - `COMPRESS_MIN_LENGTH` -- The minimal length of the session data to try compress. Defaults to 400.
    - `LOG_KEY_ERROR` -- Whether log key error caused by bad key or misconfigured `HASH_KEY_GENERATOR` or `KEY_GENERATOR`. Defaults to False.

Examples:

::

    # default: distributes sessions to 65536 Redis Hashes,
    #          use binary string form of session_key,
    #          applys Redis.randomkey 100 times to sample hash keys
    #           to check for deleting expired session,
    #          try to compress data (length >= 400) by using snappy. 
    REDIS_SESSION_CONFIG = {
        'SERVER': {},
        'USE_HASH': True,
        'KEY_GENERATOR': lambda x: x.decode('hex'),
        'HASH_KEY_GENERATOR': lambda x: x[:4].decode('hex'),
        'HASH_KEYS_CHECK_FOR_EXPIRY': lambda r: (reduce(lambda p,y :p.randomkey(),
            xrange(100), r.pipeline()).execute()),
        'COMPRESS_LIB': 'snappy',
        'COMPRESS_MIN_LENGTH': 400,
        'LOG_KEY_ERROR': False
    }
    
::

    # use session thru Redis get/set. store session key just as it is, hex string.
    REDIS_SESSION_CONFIG = {
        'USE_HASH': False,
        'KEY_GENERATOR': lambda x: x
    }

::

    # almost as above, but add key prefix 's:' for Redis keys
    REDIS_SESSION_CONFIG = {
        'USE_HASH': False,
        'KEY_GENERATOR': lambda x: 's:'+x
    }

::

    # use Redis connection instance through redisession.helper.get_redis
    # here django-redisession will try to use Redis connection instance 'foo'
    REDIS_SESSION_CONFIG = {
        'SERVER': 'foo'
    }

    # see following
    REDIS_CONFIG = {
        'default': {'db':1},
        'foo': {'db':2, unix_socket_path='/tmp/bar'}
    }

::

    # Here are some other possible settings of 'HASH_KEYS_CHECK_FOR_EXPIRY'

    # Fetching and scaning ALL keys and ALL of their fields. Use with caution. 
    lambda r: r.keys()

    # Sampling 100 hash keys out of 65536 possible slots.
    # This is useful if the number of actually used slots is closer to 65536.
    lambda r: (map(lambda x:('%04x'%x).decode('hex'),
        random.sample(xrange(65536), 100)))



You could use ``redisession.helper.get_redis`` to create and get global Redis connection instance by name. First, setting `REDIS_CONFIG` in the settings file, which is similar to DBs settings in django. For example:

::

    REDIS_CONFIG = {
        # 'name': arguments passed to redis.Redis to build a connection instance, as dict items
        'default': {'port':63790, 'db':1},
        'foo': {'db':2, unix_socket_path='/tmp/bar'},
        'session': {'db':3},
    }

then

::

    >>> from redisession.helper import get_redis
    >>> r = get_redis() # get Redis connection instance of name 'default'
    >>> r = get_redis('foo') # or of name 'foo'
    >>> r = get_redis('session') # use db 3 for sessions only
    >>> r.info()

Security
========
Isolate the Redis server of storing sessions from other usages is RECOMMENDED.
You could achieve this by using a seperate db, like conf 'session' in REDIS_CONFIG above,
or by specifying a unique key prefix in KEY_GENERATOR.
For possible security issue, refs https://www.djangoproject.com/weblog/2011/sep/09/security-releases-issued/
and https://code.djangoproject.com/changeset/16759


Cleanup expired sessions in hash mode
=====================================
    run django command `cleanuprs` in shell or cronjob.

Test
====
    ``python manager.py test redisession``. It uses your REDIS_SESSION_CONFIG settings for tests, so corresponding Redis server should be available.


CHANGES
* 0.3
    * Supporting Django 1.4
    * Fixed tests for Django 1.3 and Django 1.4
    * Declare no support for Django < 1.3
    * Added LOG_KEY_ERROR setting
    * Partially backport changesets of session from Django 1.4 for Django 1.3
    * Invalid keys for key generators no longer hit DB
* 0.2
    * Added 'cleanuprs' management command and config option
      'HASH_KEYS_CHECK_FOR_EXPIRY' to clean up session data
      in hash mode.
    * backward incompatibility: hash mode now stores expiry date in 
      big-endian order for simpler comparation, still 4 bytes.
      Please cleanup session if you're using hash mode in 0.1.
* 0.1
    * Initial version.


redisession/tests.py
from django.conf import settings
from django.contrib.sessions.tests import SessionTestsMixin
from django.utils import unittest

conf = getattr(settings, 'REDIS_SESSION_CONFIG', {})

class RedisSessionTests(SessionTestsMixin, unittest.TestCase):
    override_config = {
        'USE_HASH': False
    }
    def setUp(self):
        test_conf = conf.copy()
        test_conf.update(self.override_config)
        settings.REDIS_SESSION_CONFIG = test_conf
        from redisession.backend import SessionStore
        self.backend = SessionStore
        self.session = self.backend()

    def test_decode_django12(self):
        "We don't support Django 1.2 ever."

    def test_delete(self):
        self.session.save()
        self.session.delete(self.session.session_key)
        self.assertFalse(self.session.exists(self.session.session_key))

    if not hasattr(SessionTestsMixin, 'test_session_key_is_read_only'):
        def test_session_key_is_read_only(self):
            def set_session_key(session):
                session.session_key = session._get_new_session_key()
            self.assertRaises(AttributeError, set_session_key, self.session)


class RedisHashSessionTests(RedisSessionTests):
    override_config = {
        'USE_HASH': True
    }


redisession/backend.py
"""
A redis backend for django session, support string and hash mode.
"""

import struct
import time
from django.conf import settings
from django.contrib.sessions.backends.base import pickle, CreateError, SessionBase

conf = {
    'SERVER': {},
    'USE_HASH': True,
    'KEY_GENERATOR': lambda x: x.decode('hex'),
    'HASH_KEY_GENERATOR': lambda x: x[:4].decode('hex'),
    'HASH_KEYS_CHECK_FOR_EXPIRY': lambda r: (reduce(lambda p,y :p.randomkey(),
        xrange(100), r.pipeline()).execute()),
    'COMPRESS_LIB': 'snappy',
    'COMPRESS_MIN_LENGTH': 400,
    'LOG_KEY_ERROR': False
}
# For session key contains '0-9a-z' in incoming Django 1.5
# conf['KEY_GENERATOR'] = lambda x: x.decode('base64')
# conf['HASH_KEY_GENERATOR'] = lambda x: x.decode('base64')[:2]
conf.update(getattr(settings, 'REDIS_SESSION_CONFIG', {}))

if conf['LOG_KEY_ERROR']:
    import logging
    logger = logging.getLogger('redisession')
    logger.addHandler(logging.StreamHandler())
    logger.setLevel(logging.WARNING)

if isinstance(conf['SERVER'], dict):
    class GetRedis(object):
        def __call__(self, conf):
            if not hasattr(self, '_redis'):
                import redis
                self._redis = redis.Redis(**conf)
            return self._redis
    get_redis = GetRedis()
else:
    from redisession.helper import get_redis

if conf['COMPRESS_LIB']:
    from django.utils.importlib import import_module
    compress_lib = import_module(conf['COMPRESS_LIB'])

# TODO: flag for security verification?
FLAG_COMPRESSED = 1

class SessionStore(SessionBase):
    def __init__(self, session_key=None):
        self._redis = get_redis(conf['SERVER'])
        super(SessionStore, self).__init__(session_key)

    # XXX Try to partially comply w/ session API of newer Django (>= 1.4) for Django 1.3
    # Instead of checking Django version, test the existence directly.
    if not hasattr(SessionBase, '_get_or_create_session_key'):
        session_key = property(SessionBase._get_session_key)

        def _get_or_create_session_key(self):
            if self._session_key is None:
                self._session_key = self._get_new_session_key()
            return self._session_key

    def encode(self, session_dict):
        data = pickle.dumps(session_dict, pickle.HIGHEST_PROTOCOL)
        flag = 0
        if conf['COMPRESS_LIB'] and len(data) >= conf['COMPRESS_MIN_LENGTH']:
            compressed = compress_lib.compress(data)
            if len(compressed) < len(data):
                flag |= FLAG_COMPRESSED
                data = compressed
        return chr(flag) + data

    def decode(self, session_data):
        flag, data = ord(session_data[:1]), session_data[1:]
        if flag & FLAG_COMPRESSED:
            if conf['COMPRESS_LIB']:
                return pickle.loads(compress_lib.decompress(data))
            raise ValueError('redisession: found compressed data without COMPRESS_LIB specified.')
        return pickle.loads(data)

    def create(self):
        for i in xrange(10000):
            self._session_key = self._get_new_session_key()
            try:
                self.save(must_create=True)
            except CreateError:
                continue
            self.modified = True
            return
        raise RuntimeError('Unable to create a new session key.')

    if conf['USE_HASH']:
        def _make_key(self, session_key):
            try:
                return (conf['HASH_KEY_GENERATOR'](session_key), conf['KEY_GENERATOR'](session_key))
            except:
                if conf['LOG_KEY_ERROR']:
                    logger.warning('misconfigured key-generator or bad key "%s"' % session_key)

        def save(self, must_create=False):
            if must_create:
                func = self._redis.hsetnx
            else:
                func = self._redis.hset
            session_data = self.encode(self._get_session(no_load=must_create))
            expire_date = struct.pack('>I', int(time.time()+self.get_expiry_age()))
            key = self._make_key(self._get_or_create_session_key())
            if key is None:
                # XXX must_create = True w/ bad key or misconfigured KEY_GENERATOR,
                # which has already been logged in _make_key.
                raise CreateError
            result = func(*key, value=expire_date+session_data)
            if must_create and not result:
                raise CreateError

        def load(self):
            key = self._make_key(self._get_or_create_session_key())
            if key is not None:
                session_data = self._redis.hget(*key)
                if session_data is not None:
                    expire_date = struct.unpack('>I', session_data[:4])[0]
                    if expire_date > time.time():
                        return self.decode(session_data[4:])
            self.create()
            return {}

        def exists(self, session_key):
            key = self._make_key(session_key)
            if key is not None:
                return self._redis.hexists(*key)
            return False

        def delete(self, session_key=None):
            if session_key is None:
                if self.session_key is None:
                    return
                session_key = self.session_key
            key = self._make_key(session_key)
            if key is not None:
                self._redis.hdel(*key)

    else: # not conf['USE_HASH']
        def _make_key(self, session_key):
            try:
                return conf['KEY_GENERATOR'](session_key)
            except:
                if conf['LOG_KEY_ERROR']:
                    logger.warning('misconfigured key-generator or bad key "%s"' % session_key)

        def save(self, must_create=False):
            pipe = self._redis.pipeline()
            if must_create:
                pipe = pipe.setnx
            else:
                pipe = pipe.set
            session_data = self.encode(self._get_session(no_load=must_create))
            key = self._make_key(self._get_or_create_session_key())
            if key is None:
                # XXX must_create = True w/ bad key or misconfigured KEY_GENERATOR,
                # which has already been logged in _make_key.
                raise CreateError
            result = pipe(key, session_data).expire(key, self.get_expiry_age()).execute()
            if must_create and not (result[0] and result[1]): # for Python 2.4 (Django 1.3)
                raise CreateError

        def load(self):
            key = self._make_key(self._get_or_create_session_key())
            if key is not None:
                session_data = self._redis.get(key)
                if session_data is not None:
                    return self.decode(session_data)
            self.create()
            return {}

        def exists(self, session_key):
            key = self._make_key(session_key)
            if key is not None:
                return key in self._redis
            return False

        def delete(self, session_key=None):
            if session_key is None:
                if self.session_key is None:
                    return
                session_key = self.session_key
            key = self._make_key(session_key)
            if key is not None:
                self._redis.delete(key)
